### PR TITLE
Make sure Automatic Download downloads updates that have already been notifiied

### LINKED ISSUE #4012: Automatic download doesn't work for 'previously downloaded form only' after refusing to update manually
#### Software and hardware versions 
Collect v1.28.x, Android v all, device used...

#### Problem description
When testing #3992 I have noticed a regression issue for previously downloaded forms only option. When you select that option you will get notification about available form update and if you do not download it and change to automatic download - automatic download doesn't work even if you see that there is form uptade info in Get Blank Forms. Collect stops to recognize there is form update after refusing to download the first try. 

#### Steps to reproduce the problem
1. Download form from server versions of the form (i have tried for Aggregate)
2. Log to another server where you have version2/ upload version2 to the same server (it doesn't matter which way you choose it's reproducible for both)
3. This is an update to a form you have info that should be visible in get blank form.
4. Go to General Settings/Form Management
5. Choose Previously downloaded forms only  
6. You will get a notification immediately
7. Ignore available update 
8. Select Automatic download checkbox
9. Wait for notification and form update

You will never get notification and form won't be updated.

#### Expected behavior
After 15 minutes, notification about automatic download should display, form should be updated automatically.

#### Other information 
I have tried the same case for 1.27.3 store version and it works properly.

@getodk-bot label "regression"

### PR DESCRIPTION
**Pull request by seadowg**
Closes #4012

#### What has been done to verify that this works as intended?

Added tests to capture the edge case

#### Why is this the best possible solution? Were any other approaches considered?

Comments inline

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Good to test Previously Downloaded Only with and without Automatic Download enabled to double-check everything is the way it should be.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/database/FormDatabaseMigrator.java [hunk 1]
```diff
@@ -45,97 +24,51 @@
...
 
-/**
- * This class helps open, create, and upgrade the database file.
- */
-public class FormsDatabaseHelper extends SQLiteOpenHelper {
-    private static final String DATABASE_NAME = "forms.db";
-    public static final String FORMS_TABLE_NAME = "forms";
-
-    public static final int DATABASE_VERSION = 9;
+public class FormDatabaseMigrator {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/database/FormDatabaseMigrator.java:28]: I wanted to pull this out so we could have an object that was easier to setup and work with than the `SQLiteOpenHelper` implementations that need `DatabaseContext`.

File: collect_app/src/main/java/org/odk/collect/android/notifications/NotificationManagerNotifier.java [hunk 2]
```diff
@@ -35,18 +42,26 @@
...
+    public NotificationManagerNotifier(Application application, PreferencesProvider preferencesProvider) {
         this.application = application;
         notificationManager = (NotificationManager) application.getSystemService(NOTIFICATION_SERVICE);
+        this.preferencesProvider = preferencesProvider;
     }
 
     @Override
-    public void onUpdatesAvailable() {
+    public void onUpdatesAvailable(List<ServerFormDetails> updates) {
+        SharedPreferences metaPrefs = preferencesProvider.getMetaSharedPreferences();
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/notifications/NotificationManagerNotifier.java:59]: This is the key change here: switch away from using the database to record the versions the user is notified about and also making it more of a UI's (the `Notifier`) concern.

File: collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.JRCACHE_FILE_PATH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.JR_FORM_ID;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.JR_VERSION;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.LANGUAGE;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.MD5_HASH;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns.SUBMISSION_URI;
+import static org.odk.collect.android.provider.FormsProviderAPI.FormsColumns._ID;
+
+@RunWith(AndroidJUnit4.class)
+public class FormDatabaseMigratorTest {
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java:40]: There were tests for the database helper checking migrations, but they were out of date and `Ignore`d. Interested to know if people think we need more coverage here.

File: collect_app/src/main/java/org/odk/collect/android/database/FormDatabaseMigrator.java [hunk 4]
```diff
@@ -45,97 +24,51 @@
...
+                upgradeToVersion4(db, oldVersion);
+            case 4:
+                upgradeToVersion5(db);
+            case 5:
+                upgradeToVersion6(db);
+            case 6:
+                upgradeToVersion7(db);
+            case 7:
+                upgradeToVersion8(db);
+                break;
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/database/FormDatabaseMigrator.java]: Move `break` down to allow upgrades from versions other than 8.

File: collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java (hunk 3 above)
Comment (lognaturel) [reply collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java:40]: I know you've been trying to kill those tests for a long time. 😂 They were added after we had several significant bugs related to `onDowngrade`. I think the structure and test are good but that it would be good to also downgrade from a fictitious high version with an extra and missing column to avoid regressions there. We can assume we wouldn't touch `onDowngrade` anymore but it's hard to reason about and tempting to try to improve.

**Review by lognaturel — CHANGES_REQUESTED** [review]

File: collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java (hunk 3 above)
Comment (seadowg) [reply collect_app/src/test/java/org/odk/collect/android/database/FormDatabaseMigratorTest.java:40]: haha yeah totally happy to add more. The old tests were failing and needed rewritten anyway. I'll add a couple of downgrade tests (one with extra and one with missing columns).

File: collect_app/src/main/java/org/odk/collect/android/database/FormDatabaseMigrator.java (hunk 4 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/database/FormDatabaseMigrator.java]: Oh damn. Yes. I might try and add some tests to drive this out.

**Comment by seadowg** [comment]
@lognaturel test issues fixed. Wasn't going to be working today as usual but couldn't get this out of my head!

**Review by lognaturel — APPROVED** [review]
Yay for cleaning up unfinished business!

**Comment by kkrawczyk123** [comment]
@seadowg I have a question. Scenario from the issue seems to be fixed but I am wondering if you ignore notification about the update and do not change anything in settings - shouldn't you see the notification about an update again after 15 minutes? I think that it should appear.

**Comment by seadowg** [comment]
@kkrawczyk123 I believe the behaviour in this PR will match master though right? I do see where you're coming from: right now if you dismiss the notification it will never be seen again. I'm not sure if this is bad or not so it might be something to discuss in an issue or forum post. Apologies if I'm totally wrong and there is a regression here though!

**Comment by kkrawczyk123** [comment]
ok, @seadowg I checked on older versions and you are right. It's always good to double-check.

**Tested with success!**
Verified on Androids: 5.1, 7.1 and 10. 

Tested scenarios:
- automatic download
- match exactly
- previously downloaded forms
- scenario from issue
- used servers: Drive, Aggregate, Central
- auto send
- auto delete
- check on fill submission, match exactly and soft delete

**Comment by mmarciniak90** [comment]
It has also been successfully tested on Android versions: 7.0, 8.1, 9.0
Verified cases work as expected.

@getodk-bot unlabel "needs testing"
@getodk-bot label "behavior verified"
